Fix audio/video issues when refreshrate switch is on

a1rwulf opened the pull request:
InitModeChange is only called when the system supports the HDMI_AUDIO_PLUG event.
If the user does not configure a delay after refresh rate switching, we rely on this event to signal ResetDisplay.

If we would still start the 2s timer and the AV equipment takes longer to do the refreshrate switch, we might send ResetDisplay (and thus reconfigure AE) exactly when HDMI is unavailable which breaks the whole av pipeline.

If the user sets a delay after refreshrate we keep the old logic to start a timer with at least 2s delay, so it's up to the user to set a timer high enough to make sure HDMI is already available when the timer expires.

## Motivation and context
Fix refreshrate switch breaking A/V for subsequent playbacks.
Most of the time I had to reboot my shield to get back to a workable state.

## How has this been tested?
Nvidia Shield with maybe a shitty TV/AVR combination where I sometimes get multiple HDMI_AUDIO_PLUG events with
HDMI off and the on event takes way longer then 2sec to arrive.

## What is the effect on users?
Keep a functional kodi when refreshrate switches take longer than 2s.

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

fritsch commented on xbmc/windowing/android/WinSystemAndroid.cpp:
```diff
@@ -224,10 +224,15 @@ void CWinSystemAndroid::InitiateModeChange()
 
-  if (delay < 2000ms)
+  if (delay > 0 && delay < 2000ms)
     delay = 2000ms;
```
Might also be applicable with a GUI change, so that user knows the minimum is either 0 or 2 seconds?
The 2 seconds is only a best guess I think - is it enough / any other even in the system we could rely on?

a1rwulf replied:
Yeah I don't know if @peak3d will be able to respond what the reason for this "minimum 2s" timer thing was.
I kinda get the idea (I think): In case a HDMI_PLUG_EVENT with HDMI: on happened, https://github.com/xbmc/xbmc/blob/master/xbmc/windowing/android/WinSystemAndroid.cpp#L248 checks if the elapsed time is bigger than the "real" set delay of refreshrate and cancles the timer.
If no HDMI_AUDIO_PLUG event appears the timer will fire OnResetDisplay at least after 2s no matter what.

While it probably works for semi-broken boxes where the HDMI_AUDIO_PLUG event is not sent by the OS, it certainly breaks systems where the event works, but the connected AV equipment might take longer than 2s (which might also be rare).

Not sure if there is a need to reflect this 2s minimum in the GUI given above logic.

I tried to keep as much of the logic intact, if my change breaks tons of chinese boxes users have the chance to set delay after refershrate to 100ms which will more or less restore the original behavior.

fritsch replied:
The reason was the following: We might wait for the PLUG Event forever, it might just never come and that this won't happen there is a timeout additionally.

It was once meant to work like on e.g. X11 where one gets an even on unplug and another one on replug - but as said if the second won't come, we would hang there.

fritsch replied:
Additionally, the code logic wanted to be like: If user does not have the feature enabled, trust on the normal even only, but timeout for 2 seconds if event does not come else use whatever user told to do.

So - I think we won't find a timeout that works for everyone. As the default should work though (which is no delay set), just increase the timeout to 5 seconds if delay was not userset? With that chances are better that "out of the box" experience works.

fritsch replied:
Maybe - doing it that way:
- If user has set a timeout -> use this timeout, if hw plug comes first, wait still
- If user has set no timeout, wait for the plug event, but use a large enough timeout, so that we have a chance it works out of the box without annoying the user as user has no idea that we "workaround" his broken setup behind this back?

Keep in mind: There is no perfect solution :-) - so whatever you think is best.

a1rwulf replied:
Thx for your clarification and the ideas how to do it.
I will think a bit about it, maybe even leave it as is apparently nobody besides me has seen this issue so far...
I assume if it would be common, we would at least see a bunch of github issues about it.

fritsch replied:
Just do it like you think is right :-)

Back at the time, peak had a FireTV Cube or old FireTV or something which was semi-broken and he did not want it to hang :P. 

You care, you decide!

a1rwulf replied:
The thing is, you can call my test equipment semi-broken as it manages to have HDMI in an off state 2s after the modeset has been done.
So it's hard to argue not sending the event is considered to be "more" broken :-D
I'm really undecided at this point, needs more time and thought definitely.

fritsch replied:
How does your system behave exactly? 

You get the event properly, but **still** need to wait an additional 2 seconds after this event? And the current implementation just "ignores" your setting as the ON comes earlier?

Then I would change this implementation as user setting should always win, beside maybe one difference. We should wait for a certain period of time, when user has it off, as on "proper device" nothing will work until this event comes.

For your device it should behave like this:
User said: 5 seconds
Event comes after 1 - still wait 4.
Event does not come -> return after 5?

a1rwulf replied:
50% of the time my system works fine just like any other system with following sequence:
1. Playback start
2. Mode get's set
3. onDisplayChange arrives
4. HDMI_AUDIO_PLUG off occurs
5. HDMI_AUDIO_PLUG on occurs

All happens within those 2s and we're fine.

However the other times it goes like:
1. Playback start
2. Mode gets set
3. onDisplayChange arrives (this might happen multiple times for no apparent reason)
4. The 2s timer expires
5. OnResetDislpay gets called
6.  HDMI_AUDIO_PLUG off occurs before AE is initialized
7. AE initializes sink while HDMI is off
8. HDMI_AUDIO_PLUG on occurs at some point
9. Playback starts, I get Timeout in AddData of Audio MediaCodec
10. Playback gets stuck
11. No future playback is possible, because I always end up getting the MediaCodec audio timeout even if I kill the app -> I have to reboot to get it fixed

a1rwulf replied:
Additional notice:
I also had occasions where I got HDMI OFF, HDMI ON multiple times, but guess there's really the AVR/TV combination to blame.

fritsch replied:
I think I have the same issue on my FireTV 4K when I go through my test video files, where each one changes refreshrate / audio codec. As said before I think here the "user given setting" should in any case be taken into account.

So for your setup, two implementations are maybe a solution
a) Always wait for HDMI ON, but timeout at 5 seconds.
Pro:
- Setups that take sometimes a bit longer now always work. But sometimes you wait 3, and otherwise 2 or max 5 seconds
Cons:
- Semi-broken setups would always wait 5 seconds
b) Always wait at  least the time the user sets in the setting
Pro:
- You set it to 5 seconds and it will always behave the same
Cons:
- It will always need 5 seconds

For jump off / jump on / AE needed to react maybe faster, but as you said -> box needs reboot :-(

From a gut feeling, while I think option a) would be better for your setup. I think option b) is better for a user, cause they can decide by themselves ...

fritsch replied:
```
    if (m_dispResetTimer->IsRunning())
    {
      // We stop the timer if OS supports HDMI_AUDIO_PLUG intent
      // and configured delay is smaller than the time HDMI_PLUG took.
      // Note that timer is always started with minimum of 2 seconds
      // regardless if the configured delay is smaller
      if (m_dispResetTimer->GetElapsedMilliseconds() >=
          CServiceBroker::GetSettingsComponent()->GetSettings()->GetInt(
              "videoscreen.delayrefreshchange") *
              100)
        m_dispResetTimer->Stop();
      else
        return; // <- This will cause the RenderIMPl to do this job
    }

    for (auto resource : m_resources)
      resource->OnResetDisplay(); // <- this is only triggered from here when the display ON comes back and timer is gone. Else the timer would timeout at a point and it is triggered from: WinSystemAndroidGLESContext.cpp
    m_dispResetTimer->Stop();
    m_HdmiModeTriggered = false;
  }
```

I am wondering if this issue is actually force triggered xbmc/windowing/android/WinSystemAndroidGLESContext.cpp which causes nothing to wait, whenever no timer is set?

It seems the timer is once set from XBMCApp.cpp whenever the system gets the PLUG event, then it is checked if the "user set" timer is already gone, if not - it just returns. Then the RenderIMpl from above always checks if the timer is still runing (per frame) and tries to wakeup the resources.

So, the 2 seconds is actually a workaround for RenderIMpl - but I think your setup should already work right now - if you set it to 5 seconds? Right?

a1rwulf replied:
Yeah I assume 5s will be fine for my setup, unless I stumble across runs where it takes even longer.
Will do some more testing and collect some logs of fail runs.

a1rwulf committed:
Fix audio/video issues when refreshrate switch is on

InitModeChange is only called when the system supports
the HDMI_AUDIO_PLUG event.
If the user does not configure a delay after refresh rate
switching, we rely on this event to signal ResetDisplay.

If we would still start the 2s timer and the AV equipment takes
longer to do the refreshrate switch, we might send
ResetDisplay (and thus reconfigure AE) exactly when HDMI
is unavailable which breaks the whole av pipeline.

If the user sets a delay after refreshrate we keep the old logic
to start a timer with at least 2s delay, so it's up to the
user to set a timer high enough to make sure HDMI is already
available when the timer expires.

CastagnaIT commented:
@a1rwulf please can you rebase with current master? this PR cannot be builded

a1rwulf commented:
@CastagnaIT I think it was because of a mistake in my code, not because of a conflict.
Rebased anyway, let's see if it works now.

a1rwulf committed:
fixup! Fix audio/video issues when refreshrate switch is on

a1rwulf commented:
Builds fine locally, not sure what's wrong atm.
Jenkins build this please.

fritsch commented:
@a1rwulf is it working for you?

a1rwulf commented:
Yeah current state works for me and I think also for @CastagnaIT , but I haven't had time last week to work on it and do more testing.

fritsch commented:
Okay. The issue I see here now is: Devices which don't properly send the hdmi event would basically never wakeup, right? For this the "default" 2 second timer was used.

If that's the case, maybe: Off, Auto and the time selection might be a gui indication. Off being the default. For Auto one can write, that there might be an issue, that devices never wakeup or combine it with a 20 second timer.

What do you think?
I don't want to break "broken firmware" Android devices, which actually are the most ... broken to a certain regard.

a1rwulf commented:
@fritsch I have no plans to merge this as is - I totally understand the implications and I'm not going to break tons of existing devices for no reason.
Unfortunately I had zero time last 2 weeks to think about how I want to go ahead with this.
I like all of your suggestions, but I need more time to implement a few variations and do tests on several of my devices.
Next step for me is to work through all my cheap boxes to see if I own some "broken" ones too, which will make it easier to work on a solution that works for me with decent and "broken" firmwares.

fritsch commented:
@a1rwulf Could you rebase a final time? I want to test this on my FireTV Cube / 4Ks - as when I was looking into the details yesterdays ... it's a whole lot of a mess in our current streams.

With my change recently to retry ... one can see now for the first time, that things don't work properly.
